sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala

```diff
@@ -89,15 +87,18 @@ case class ColumnarToRowExec(child: SparkPlan) extends ColumnarToRowTransition w
   override def doExecute(): RDD[InternalRow] = {
     val numOutputRows = longMetric("numOutputRows")
     val numInputBatches = longMetric("numInputBatches")
-    // This avoids calling `output` in the RDD closure, so that we don't need to include the entire
-    // plan (this) in the closure.
-    val localOutput = this.output
-    child.executeColumnar().mapPartitionsInternal { batches =>
-      val toUnsafe = UnsafeProjection.create(localOutput, localOutput)
-      batches.flatMap { batch =>
-        numInputBatches += 1
-        numOutputRows += batch.numRows()
-        batch.rowIterator().asScala.map(toUnsafe)
+    val evaluatorFactory =
+      new ColumnarToRowEvaluatorFactory(
+        child.output,
+        numOutputRows,
+        numInputBatches)
+
+    if (conf.usePartitionEvaluator) {
+      child.executeColumnar().mapPartitionsWithEvaluator(evaluatorFactory)
+    } else {
+      child.executeColumnar().mapPartitionsInternal { batches =>
+        val evaluator = evaluatorFactory.createEvaluator()
+        evaluator.eval(0, batches)
```

Comment:
We don't need pass real partition index?

Reply:
In the original code, the index was not used as `mapPartitionsInternal` is called

Reply:
This is not right. Even if it's not used for now, we should still set it correctly to be future-proof.

Reply:
I'm fixing it at https://github.com/apache/spark/pull/42185

Reply:
@cloud-fan , Thanks for the fix,
I can fix similar issues in other merged PRs
https://github.com/apache/spark/pull/42024/files
https://github.com/apache/spark/pull/42025/files
https://github.com/apache/spark/pull/41884/files